libraries/plugins/xfpga/usrclk/fpga_user_clk.c

```diff
@@ -0,0 +1,816 @@
+		(*seq)++);
+	usleep(IOPLL_CAL_DELAY_MS);
+	return res;
+}
+
+fpga_result get_usrclk_uio(const char *sysfs_path,
+	uint32_t feature_id,
+	struct opae_uio *uio,
+	uint8_t **uio_ptr)
+{
+	fpga_result res                   = FPGA_NOT_FOUND;
+	char feature_path[SYSFS_PATH_MAX] = { 0 };
+	char dfl_dev_str[SYSFS_PATH_MAX]  = { 0 };
+	int gres, ret                     = 0;
+	size_t i                          = 0;
+	uint64_t value                    = 0;
+	glob_t pglob;
+
+	if (sysfs_path == NULL) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	if (snprintf(feature_path, sizeof(feature_path),
+		"%s/dfl_dev*/feature_id",
+		sysfs_path) < 0) {
+		OPAE_ERR("snprintf buffer overflow");
+		return FPGA_EXCEPTION;
+	}
+
+	gres = glob(feature_path, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		OPAE_ERR("Failed pattern match %s: %s",
+			feature_path, strerror(errno));
+		if (pglob.gl_pathv)
+			globfree(&pglob);
+		return FPGA_INVALID_PARAM;
+	}
+
+	for (i = 0; i < pglob.gl_pathc; i++) {
+		res = sysfs_read_u64(pglob.gl_pathv[i], &value);
+		if (res != FPGA_OK) {
+			OPAE_MSG("Failed to read sysfs value");
+			continue;
+		}
+
+		if (value == feature_id) {
+			res = FPGA_OK;
+			char *p = strstr(pglob.gl_pathv[i], "dfl_dev");
+			if (p == NULL) {
+				res = FPGA_NOT_FOUND;
+				goto free;
+			}
+
+			char *end = strchr(p, '/');
+			if (end == NULL) {
+				res = FPGA_NOT_FOUND;
+				goto free;
+			}
+			strncpy(dfl_dev_str, p, end - p);
+			*(dfl_dev_str + (end - p)) = '\0';
+
+			ret = opae_uio_open(uio, dfl_dev_str);
+			if (ret) {
+				OPAE_ERR("Failed to open uio");
+				break;
+			}
+
+			ret = opae_uio_region_get(uio, 0, (uint8_t **)uio_ptr, NULL);
+			if (ret) {
+				OPAE_ERR("Failed to get uio region");
+				opae_uio_close(uio);
+				break;
+			}
+		}
+		break;
+	}
+
+free:
+	if (pglob.gl_pathv)
+		globfree(&pglob);
+	return res;
+}
+
+//Get fpga user clock
+fpga_result get_userclock(const char *sysfs_path,
+	uint64_t *userclk_high,
+	uint64_t *userclk_low)
+{
+	char sysfs_usrpath[SYSFS_PATH_MAX]  = {0};
+	fpga_result result                  = FPGA_OK;
+	uint32_t high, low                  = 0;
+	int ret                             = 0;
+	uint8_t *uio_ptr                    = NULL;
+	struct opae_uio uio;
+
+	memset(&uio, 0, sizeof(uio));
+
+	if ((sysfs_path == NULL) ||
+		(userclk_high == NULL) ||
+		(userclk_low == NULL)) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	// Test for the existence of the userclk_frequency file
+	// which indicates an S10 driver
+	ret = using_iopll(sysfs_usrpath, sysfs_path);
+	if (ret == FPGA_OK) {
+		result = sysfs_read_u32_pair(sysfs_usrpath, &low, &high, ' ');
+		if (FPGA_OK != result)
+			return result;
+
+		*userclk_high = high * 1000;	// Adjust to Hz
+		*userclk_low = low * 1000;
+		return FPGA_OK;
+	} else if (ret == FPGA_NO_ACCESS) {
+		return FPGA_NO_ACCESS;
+	}
+
+	result = get_usrclk_uio(sysfs_path,
+		0x14,
```

Comment:
I think this is the feature id.  Can we use a #define for it?

Reply:
fixed  added  #define USRCLK_FEATURE_ID 0x14